### PR TITLE
Updates for Python 3.12 compatibility

### PR DESCRIPTION
**Pull request by jtniehof**
This PR is the "get everything working for 3.12" branch.

- [x] Squash existing deprecations on up-to-date 3.10
- [x] Update CI, test scripts, build scripts, requirements for 3.11 and squash any errors/deprecations. This takes care of #717 but I'm not going to magic-word it until the release is out, so people can still find it in the meantime.
- [x] Update CI, test scripts, build scripts, requirements for 3.12 and squash any errors/deprecations
- [x] Try to future-proof for 3.13 (e.g. look at what minimum numpy might be) (EDIT: Deferred to #725 which will remove minimum numpy)
- [x] Quick scrub for pre-3.6 code and old spacepy deprecations
- [x] Try to make the CI fail on deprecation warnings (Deferred to #721)
- [x] Update to CDF 3.9 while we're at it
- [x] namespace packages warnings

Other changes:

- coordinates and irbem started warning about backend changes with 0.3.0, default changed to SpacePy in 0.4.0 but the warning stayed. The warning has been removed (and a bunch of code hiding it when not relevant).
- In time, removed some warnings for defaults that changed in 0.2.2, and removed a little-used argument that was deprecated in 0.2.2.
- In pybats, removed `bats.add_b_magsphere_new` and `bats.add_b_magsphere_legacy` (deprecated in 0.2.0)

## PR Checklist

- [X] Pull request has descriptive title
- [X] Pull request gives overview of changes
- [x] New code has inline comments where necessary
- [x] (N/A) Any new modules, functions or classes have docstrings consistent with SpacePy style
- [x] Added an entry to release notes if fixing a significant bug or providing a new feature
- [x] (N/A) New features and bug fixes should have unit tests
- [x] Relevant issues are linked to (e.g. `See issue #` or `Closes #`)

### CODE REVIEW

**Comment by jtniehof** [comment]
Current CI failure on 3.12/latest numpy is numpy/numpy#24838, hopefully can work around it but the upshot might be requiring meson to build. In that case we might have to set a max numpy version in build-requires and just get 0.5.0 out, then work on switching from f2py to ctypes for irbem.

**Comment by jtniehof** [comment]
Yep, f2py requires meson now, will have to deliver 0.5.0 without but should be okay to do wheels. #725.

File: spacepy/toolbox/__init__.py [hunk 1]
```diff
@@ -1433,33 +1399,33 @@ def update(all=True, QDomni=False, omni=False, omni2=False, leapsecs=False,
     return datadir
 
 def indsFromXrange(inxrange):
-    '''return the start and end indices implied by an xrange, useful when xrange is zero-length
+    '''return the start and end indices implied by an range, useful when range is zero-length
```
Comment (drsteve) [inline spacepy/toolbox/__init__.py:1402]: an -> a

File: tests/test_irbempy.py [hunk 2]
```diff
@@ -253,20 +236,12 @@ def test_get_dtype(self):
 
     def test_prep_irbem_sysaxesnone(self):
         """prep_irbem should handle 'car' and 'sph' version of systems identically"""
-        with warnings.catch_warnings():
-            warnings.filterwarnings('ignore', message=r'Use of IRBEM to perform',
-                                    category=DeprecationWarning,
-                                    module=r'spacepy.coordinates$')
-            locc = spacepy.coordinates.Coords([[3, 0, 0], [2, 0, 0]],
+        locc = spacepy.coordinates.Coords([[3, 0, 0], [2, 0, 0]],
                                               'GSM', 'car', use_irbem=True)
```
Comment (drsteve) [inline tests/test_irbempy.py:240]: Visual alignment of indentation?

**Review by drsteve — COMMENTED** [review]
No concerns from me beyond the minor comments already posted. I have not tested the integration or sandbox stuff. I'll defer to @spacecataz on the function removal in pybats.

File: spacepy/pybats/__init__.py [hunk 3]
```diff
@@ -1714,7 +1714,7 @@ class ImfInput(PbData):
     '''
     A class to read, write, manipulate, and visualize solar wind upstream
     input files for SWMF simulations.  More about such files can be found
-    in the SWMF/BATS-R-US documentation for the \#SOLARWINDFILE command.
+    in the SWMF/BATS-R-US documentation for the \\#SOLARWINDFILE command.
```
Comment (spacecataz) [inline spacepy/pybats/__init__.py:1717]: Quick question.
In the original, the single backslash in `\#SOLARWINDFILE` was to escape `#` (I think.) Now, we're escaping `\`, which I don't think is necessary. I think we just want to remove the `\\` preceding the hash, unless I'm missing something with Sphinx...

**Review by spacecataz — APPROVED** [review]
Things look good- changes are logical and clean. Happy to see some deprecated items get chopped.

**Pull request merged**